src/sdk/pynni/nni/bohb_advisor/bohb_advisor.py

```diff
@@ -0,0 +1,575 @@
+                return None
+            this_round_perf = self.configs_perf[i]
+            if self.optimize_mode is OptimizeMode.Maximize:
+                sorted_perf = sorted(this_round_perf.items(
+                ), key=lambda kv: kv[1][1], reverse=True)  # reverse
+            else:
+                sorted_perf = sorted(
+                    this_round_perf.items(), key=lambda kv: kv[1][1])
+            logger.debug(
+                'bracket %s next round %s, sorted hyper configs: %s', self.s, self.i, sorted_perf)
+            next_n, next_r = self.get_n_r()
+            logger.debug('bracket %s next round %s, next_n=%d, next_r=%d',
+                         self.s, self.i, next_n, next_r)
+            hyper_configs = dict()
+            for k in range(next_n):
+                params_id = sorted_perf[k][0]
+                params = self.hyper_configs[i][params_id]
+                params[_KEY] = next_r  # modify r
+                # generate new id
+                increased_id = params_id.split('_')[-1]
+                new_id = create_bracket_parameter_id(
+                    self.s, self.i, increased_id)
+                hyper_configs[new_id] = params
+            self._record_hyper_configs(hyper_configs)
+            return [[key, value] for key, value in hyper_configs.items()]
+        return None
+
+    def get_hyperparameter_configurations(self, num, r, config_generator):
+        """generate num hyperparameter configurations from search space using Bayesian optimization
+
+        Parameters
+        ----------
+        num: int
+            the number of hyperparameter configurations
+
+        Returns
+        -------
+        list
+            a list of hyperparameter configurations. Format: [[key1, value1], [key2, value2], ...]
+        """
+        global _KEY
+        assert self.i == 0
+        hyperparameter_configs = dict()
+        for _ in range(num):
+            params_id = create_bracket_parameter_id(self.s, self.i)
+            params = config_generator.get_config(r)
+            params[_KEY] = r
+            hyperparameter_configs[params_id] = params
+        self._record_hyper_configs(hyperparameter_configs)
+        return [[key, value] for key, value in hyperparameter_configs.items()]
+
+    def _record_hyper_configs(self, hyper_configs):
+        """after generating one round of hyperconfigs, this function records the generated hyperconfigs,
+        creates a dict to record the performance when those hyperconifgs are running, set the number of finished configs
+        in this round to be 0, and increase the round number.
+
+        Parameters
+        ----------
+        hyper_configs: list
+            the generated hyperconfigs
+        """
+        self.hyper_configs.append(hyper_configs)
+        self.configs_perf.append(dict())
+        self.num_finished_configs.append(0)
+        self.num_configs_to_run.append(len(hyper_configs))
+        self.increase_i()
+
+
+class BOHB(MsgDispatcherBase):
```

Comment:
@QuanluZhang @PurityFan  still do not recommend directly implementing BOHB from MsgDispatcherBase. We definitely need a better abstraction for 'Advisor'. Current implementation actually couples BOHB algorithm with Tuner & Advisor logics.

Reply:
yes, will discuss this part in future :)